OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+#include <SimTKcommon.h>
+
+
+namespace OpenSim {
+  class AbstractDataTable;
+  template<typename ET> class DataTable_;
+
+  using DataTable = DataTable_<SimTK::Real>;
+
+  /// Enum to specify if the the input iterator is traversing data row-wise or
+  /// col-wise.
+  /// clang3.6 crashes if this is turned to a "enum class". Until it is fixed,
+  /// use a pre c++11 enum.
+  enum InputItDir{
+    ROWWISE, 
+    COLWISE
+  };
+
+  /// Bind two DataTables by row.
+  template<typename ET>
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
+class OpenSim::AbstractDataTable {
+public:
+  AbstractDataTable() = default;
+  AbstractDataTable(const AbstractDataTable&) = delete;
+  AbstractDataTable(AbstractDataTable&&) = delete;
+  virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+  virtual ~AbstractDataTable() {}
+}; // AbstractDataTable
+
+class OpenSim::ColumnDoesNotExist : public std::runtime_error {
+public:
+  ColumnDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasLabel : public std::runtime_error {
+public:
+  ColumnHasLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasNoLabel : public std::runtime_error {
+public:
+  ColumnHasNoLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ZeroElements : public std::runtime_error {
+public:
+  ZeroElements(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::InvalidEntry : public std::runtime_error {
+public:
+  InvalidEntry(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyExists : public std::runtime_error {
+public:
+  MetaDataKeyExists(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyDoesNotExist : public std::runtime_error {
+public:
+  MetaDataKeyDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataTypeMismatch : public std::runtime_error {
+public:
+  MetaDataTypeMismatch(const std::string& expl) : runtime_error(expl) {}
+};
```

Comment:
These exceptions are awesome to have. Any particular reason for deriving from `std::runtime_error` instead of `OpenSim::Exception`? `OpenSim::Exception` is wrapped by SWIG; I'm not sure how these exceptions would look in SWIG.